crates/runtime/src/search/rrf.rs

```diff
@@ -0,0 +1,631 @@
+    /// ...into a neat struct of subquery expressions and an optional user-provided smoothing parameter.
+    ///
+    /// # Arguments
+    /// * `args` - A slice of `Expr` containing search UDTF invocations and an optional `k` parameter
+    ///
+    /// # Returns
+    /// * `Ok(ReciprocalRankFusionArgs)` - Successfully parsed arguments
+    /// * `Err` - If fewer than 2 search queries are provided or if unparsing fails
+    pub fn from_udtf_exprs(args: &[Expr]) -> Result<ReciprocalRankFusionArgs> {
+        let mut search_udtfs: Vec<Expr> = vec![];
+        let mut k_argument: Option<f64> = None;
+        let mut join_pk_argument: Option<Expr> = None;
+
+        for expr in args {
+            match expr {
+                e @ Expr::ScalarFunction(_) => search_udtfs.push(e.clone()),
+                Expr::Literal(ScalarValue::Float64(Some(k)), ..) if k_argument.is_none() => {
+                    k_argument = Some(*k);
+                }
+                Expr::Column(c) if join_pk_argument.is_none() => {
+                    join_pk_argument = Some(col(c.name.clone()));
+                }
+                // Show a useful error for the rest
+                other_expr => {
+                    return Err(DataFusionError::NotImplemented(format!(
+                        "{RRF_UDF_NAME} does not yet support {other_expr} arguments."
+                    )));
+                }
+            }
+        }
+
+        if search_udtfs.len() < 2 {
+            return Err(DataFusionError::Plan(format!(
+                "{RRF_UDF_NAME} needs at least 2 search queries to fuse results."
+            )));
+        }
+
+        Ok(Self {
+            search_udtf_exprs: search_udtfs,
+            k: k_argument.unwrap_or(60.0),
+            join_key: join_pk_argument,
+        })
+    }
+}
+
+impl Debug for ReciprocalRankFusion {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ReciprocalRankFusion")
+    }
+}
+
+pub struct ReciprocalRankFusion {
+    pub session_context: Arc<SessionContext>,
+    df: Option<DataFrame>,
+}
+
+// TODO: DF support for nested UDTF calls without ScalarUDF "hack"
+impl ReciprocalRankFusion {
+    #[must_use]
+    pub fn from_ctx(session_context: &Arc<SessionContext>) -> Self {
+        Self {
+            session_context: Arc::clone(session_context),
+            df: None,
+        }
+    }
+
+    #[must_use]
+    pub fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    #[must_use]
+    pub fn with_df(mut self, df: DataFrame) -> Self {
+        self.df = Some(df);
+        self
+    }
+
+    fn scalar_stub_error<T>() -> Result<T, DataFusionError> {
+        exec_err!(
+            "{RRF_UDF_NAME} is a table function with a scalar stub. Please call as a table function."
+        )
+    }
+
+    fn empty_df(&self) -> DataFrame {
+        let schema = Schema::new(vec![Field::new("dummy", DataType::Int64, false)]);
+
+        let empty_batch = RecordBatch::new_empty(Arc::new(schema));
```

Comment:
Can this be just `RecordBatch::new_empty(Arc::new(Schema::empty()));`? Do we need `dummy` field/column